mkosi/__init__.py

```diff
@@ -3641,47 +3636,47 @@ def insert_verity(
         )
 
 
-PKCS7_NOCERTS  = 0x2    # Don't include signature certificates in result
-PKCS7_DETACHED = 0x40   # Don't include data to sign in result
-PKCS7_BINARY   = 0x80   # Don't mangle newlines for MIME canonical format
-PKCS7_NOATTR   = 0x100  # Don't include signature time, … in result
-
-
 def make_verity_sig(
     args: CommandLineArguments, root_hash: Optional[str], do_run_build_script: bool, for_cache: bool
 ) -> Tuple[Optional[BinaryIO], Optional[bytes], Optional[str]]:
+    from cryptography.x509 import x509
+    from cryptography.hazmat.primitives import hashes
+    from cryptography.hazmat.primitives import serialization
+    from cryptography.hazmat.primitives.serialization import pkcs7
 
     if do_run_build_script or args.verity != "signed":
         return None, None, None
     if for_cache:
         return None, None, None
 
     assert root_hash is not None
-    assert crypto is not None
 
     with complete_step("Signing verity root hash…"):
-
-        key = crypto.load_privatekey(crypto.FILETYPE_PEM, args.secure_boot_key.read_text())
-        certificate = crypto.load_certificate(crypto.FILETYPE_PEM, args.secure_boot_certificate.read_text())
-
-        # Note that the library returns the SHA256 digest in an
-        # uppecase format with : as byte separators. Let's convert to
-        # classic lowercase series of unseparated hex digits
-        fingerprint = certificate.digest("sha256").decode("ascii").replace(":", "").lower()
-
-        bio_in = crypto._new_mem_buf(root_hash.encode("utf-8"))
-        pkcs7 = crypto._lib.PKCS7_sign(
-            certificate._x509,
-            key._pkey,
-            crypto._ffi.NULL,
-            bio_in,
-            PKCS7_DETACHED | PKCS7_NOCERTS | PKCS7_NOATTR | PKCS7_BINARY,
+        key = serialization.load_pem_private_key(args.secure_boot_key.read_bytes(), password=None)
+        certificate = x509.load_pem_x509_certificate(args.secure_boot_certificate.read_bytes())
+
+        fingerprint = certificate.fingerprint(hashes.SHA256()).hex()
+
+        signature_builder = pkcs7.PKCS7SignatureBuilder()
+        signature = signature_builder.add_signer(
+            certificate,
+            key,
+            hashes.SHA256()
+        ).set_data(
+            root_hash.encode("utf-8")
+        ).sign(
+            options=[
+                pkcs7.PKCS7Options.DetachedSignature,
+                pkcs7.PKCS7Options.NoCerts,
+                pkcs7.PKCS7Options.NoAttributes,
+                pkcs7.PKCS7Options.Binary
+            ],
+            encoding=serialization.Encoding.PEM
         )
-        bio_out = crypto._new_mem_buf()
-        crypto._lib.i2d_PKCS7_bio(bio_out, pkcs7)
-        sigbytes = crypto._bio_to_string(bio_out)
 
-        b64encoded = base64.b64encode(sigbytes).decode("ascii")
+        # Remove -----BEGIN PKCS7----- and -----END PKCS7----- from beginning and end
+        b64encoded = "".join(signature.decode("ascii").splitlines()[1:-1])
```

Comment:
this probably also includes newlines? we don't want those either... i.e. i think we really should just do the base64 encoding ourselves, after all we want a base64-encoded DER, and not a PEM file if you so will, i.e. the base64 need is because that's kinda how things are done with embedding binary data in json, it's not because we actually wanted PEM...

Reply:
Yes, this includes newlines, which the `.splitlines()` and `"".join` remove.

`cryptography`'s general approach is trying to cushion off sharp edges and make all things cryptography hard to misuse. The `PKCS7SignatureBuilder()` only returns things with their valid encodings (`serialization.Encoding`) and I haven't gone looking for more primitive primitives, but just hacked this together quickly.

I did compare both the current version and this one with the same input root hash and the base64 encoded blobs came out the same.